contracts/dk/DuelistKingStaking.sol

```diff
@@ -0,0 +1,253 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.8.4 <0.9.0;
+pragma abicoder v2;
+
+import '@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol';
+import '@openzeppelin/contracts/token/ERC20/ERC20.sol';
+import '@openzeppelin/contracts/utils/Address.sol';
+import '../libraries/RegistryUser.sol';
+
+/**
+ * Duelist King Staking
+ * Name: DuelistKingStaking
+ * Doamin: Duelist King
+ */
+
+contract DuelistKingStaking is RegistryUser {
+  using SafeERC20 for ERC20;
+  using Address for address;
+
+  struct StakingCampaign {
+    uint64 startDate;
+    uint64 endDate;
+    uint64 returnRate;
+    uint64 rewardPhaseBoxId;
+    uint256 maxAmountOfToken;
+    uint256 stakedAmountOfToken;
+    uint256 limitStakingAmountForUser;
+    address tokenAddress;
+    uint64 maxNumberOfBoxes;
+  }
+
+  struct UserStakingSlot {
+    uint256 stakingAmountOfToken;
+    uint128 stakedAmountOfBoxes;
+    uint64 startStakingDate;
+    uint64 lastStakingDate;
+  }
+
+  address private _owner;
+  uint256 private _totalCampaign;
+  uint32 constant decimals = 1000000;
+
+  mapping(uint256 => StakingCampaign) private _campaignStorage;
+
+  mapping(uint256 => mapping(address => UserStakingSlot)) private _userStakingSlot;
+
+  mapping(address => uint256) private _totalPenalty;
+
+  // New created campaign event
+  event NewCampaign(uint64 indexed startDate, address indexed tokenAddress, uint256 indexed campaignId);
+
+  // Staking event
+  event Staking(address indexed owner, uint256 indexed amount, uint256 indexed startStakingDate);
+
+  // Unstaking event
+  event Unstaking(address indexed owner, uint256 indexed tokenAmount, uint64 indexed unStakeTime);
+
+  // User claim boxes event
+  event ClaimRewardBoxes(address indexed owner, uint128 indexed numberOfBoxes, uint64 indexed rewardPhaseBoxId);
+
+  event Withdrawal(address indexed beneficiary, uint256 indexed amount, uint64 indexed timestamp);
+
+  constructor(address registry_, bytes32 domain_) {
+    _registryUserInit(registry_, domain_);
+  }
+
+  /*******************************************************
+   * Public section
+   *******************************************************/
+
+  function staking(uint256 campaignId, uint256 amountOfToken) external returns (bool) {
+    StakingCampaign memory currentCampaign = _campaignStorage[campaignId];
+    UserStakingSlot memory currentUserStakingSlot = _userStakingSlot[campaignId][msg.sender];
+    ERC20 currentToken = ERC20(currentCampaign.tokenAddress);
+
+    // User should stake at least 1 day in event time
+    require(
+      block.timestamp >= currentCampaign.startDate && block.timestamp <= currentCampaign.endDate - 1 days,
```

Comment:
It should be `block.timestamp < currentCampaign.endDate - (1 days + 1 hours)`, it can hit boundary condition